src/debugger/VariableManager.js

```
@@ -0,0 +1,100 @@
+import { ReferenceMap } from "./ReferenceMap.js";
+import { LexicalEnvironment, DeclarativeEnvironmentRecord } from "./../environment.js";
+import { Value, ConcreteValue, PrimitiveValue } from "./../values/index.js";
+
+// This class manages the handling of variable requests in the debugger
+// The DebugProtocol specifies collections of variables are to be fetched using a
+// unique reference ID called a variablesReference. This class can generate new
+// variablesReferences to pass to the UI and then perform lookups for those
+// variablesReferences when they are requested.
+export class VariableManager {
+  constructor() {
+    this._containerCache = new Map();
+    this._referenceMap = new ReferenceMap();
+  }
+  // cache for created references
+  _containerCache: Map<VariableContainer, number>;
+  // map for looking up references
+  _referenceMap: ReferenceMap<VariableContainer>;
+
+  // Given a container, either returns a cached reference for that container if
+  // it exists or return a new reference
+  getReferenceForValue(value: VariableContainer): number {
+    let cachedRef = this._containerCache.get(value);
+    if (cachedRef) {
+      return cachedRef;
+    }
+
+    let varRef = this._referenceMap.add(value);
+    this._containerCache.set(value, varRef);
+    return varRef;
+  }
+
+  // The entry point for retrieving a collection of variables by a reference
+  getVariablesByReference(reference: number): Array<Variable> {
+    let container = this._referenceMap.get(reference);
+    if (!container) return [];
+    if (container instanceof LexicalEnvironment) {
+      return this._getVariablesFromEnv(container);
+    }
+    return [];
+  }
+
+  _getVariablesFromEnv(env: LexicalEnvironment): Array<Variable> {
+    let envRecord = env.environmentRecord;
+    if (envRecord instanceof DeclarativeEnvironmentRecord) {
+      return this._getVariablesFromDeclarativeEnv(envRecord);
+    }
+    return [];
```

The same, either invariant to fail or add a TODO comment instead of silently return empty collection.